src/UnityExtension/Assets/Editor/GitHub.Unity/UI/TreeControl.cs

```diff
@@ -257,26 +255,116 @@ public void Blur()
+            var isChecked = false;
+
+            switch (node.CheckState)
+            {
+                case CheckState.Mixed:
+                case CheckState.Empty:
+                    node.CheckState = CheckState.Checked;
+                    isChecked = true;
+                    break;
+
+                case CheckState.Checked:
+                    node.CheckState = CheckState.Empty;
+                    break;
+            }
+
             if (node.IsFolder)
             {
-                
+                ToggleChildrenChecked(idx, node, isChecked);
             }
-            else
+
+            ToggleParentFoldersChecked(idx, node, isChecked);
+        }
+
+        private void ToggleChildrenChecked(int idx, TreeNode node, bool isChecked)
+        {
+            for (var i = idx + 1; i < nodes.Count && node.Level <= nodes[i].Level; i++)
             {
-                switch (node.CheckState)
+                var childNode = nodes[i];
+                if (node.Level == childNode.Level)
                 {
-                    case CheckState.Empty:
-                        node.CheckState = CheckState.Checked;
-                        break;
+                    continue;
+                }
+
+                childNode.CheckState = isChecked ? CheckState.Checked : CheckState.Empty;
 
-                    case CheckState.Checked:
-                        node.CheckState = CheckState.Empty;
-                        break;
+                if (childNode.IsFolder)
+                {
+                    ToggleChildrenChecked(i, childNode, isChecked);
                 }
+            }
+        }
+
+        private void ToggleParentFoldersChecked(int idx, TreeNode node, bool isChecked)
+        {
+            while (true)
+            {
+                if (node.Level > 0)
+                {
+                    var siblingsInSameState = true;
+                    var firstSiblingIndex = idx;
+
+                    for (var i = idx - 1; i > 0 && node.Level <= nodes[i].Level; i--)
+                    {
+                        var previousNode = nodes[i];
+                        if (node.Level < previousNode.Level)
+                        {
+                            continue;
+                        }
+
+                        firstSiblingIndex = i;
+
+                        if (siblingsInSameState)
+                        {
+                            var previousNodeIsChecked = previousNode.CheckState == CheckState.Checked;
 
-                Debug.LogFormat("Ripple CheckState index:{0} level:{1}", idx, node.Level);
+                            if (isChecked != previousNodeIsChecked)
+                            {
+                                siblingsInSameState = false;
+                            }
+                        }
+                    }
+
+                    if (siblingsInSameState)
+                    {
+                        for (var i = idx + 1; i < nodes.Count && node.Level <= nodes[i].Level; i++)
```

Comment:
Same here for not being the last node? Guaranteed?

Reply:
Here we are going forwards from our current position, and the for loop bounds will keep us in the list. We are looking for the end of the list or a point where the level is higher than the current position. We should be good here.